rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3Client.ts

```diff
@@ -103,27 +124,35 @@ export class AmazonS3Client {
 
   public async getObjectAsync(objectName: string): Promise<Buffer | undefined> {
     this._writeDebugLine('Reading object from S3');
-    const response: fetch.Response = await this._makeRequestAsync('GET', objectName);
-    if (response.ok) {
-      return await response.buffer();
-    } else if (response.status === 404) {
-      return undefined;
-    } else if (response.status === 403 && !this._credentials) {
-      return undefined;
-    } else {
-      this._throwS3Error(response, await this._safeReadResponseText(response));
-    }
+    const sendRequest: () => Promise<Buffer | undefined> = async (): Promise<Buffer | undefined> => {
+      const response: fetch.Response = await this._makeRequestAsync('GET', objectName);
+      if (response.ok) {
+        return await response.buffer();
+      } else if (response.status === 404) {
+        return undefined;
+      } else if (response.status === 403 && !this._credentials) {
```

Comment:
`400` and `401` are not retriable

Reply:
I added 2 tests to verify that 400 and 401 are retriable:

https://github.com/microsoft/rushstack/pull/3248/commits/773aaccc80740fb4cb4fd6509494557b42315a97

Reply:
To clarify, I meant that they should be considered to be not retriable, since missing/incorrect auth is not something that can be corrected by sending the request again, nor is a malformed request payload.

Reply:
Ah that makes sense 😅. I'll update so that the code can either throw error or retry

Reply:
Do you know the rationale behind only throwing 403 errors if credentials are provided? This code: `if (response.status === 403 && !this._credentials) return undefined` 

Should 401 and 400 be handled in the same way?

Reply:
Also, in `uploadObjectAsync` response codes are not handled (it is just `if (!response.ok) throw err`). Do you think response codes in `uploadObjectAsync` should be handled in the same way as in `getObjectAsync`?

Reply:
To answer my question myself:

I'd say 400, 401, 403 should be handled equally.

When no credentials are provided unauth errors should be silenced as the ones potentially running into this issue usually are human users (not CI) that don't have credentials and they should still be able to build without access to the cloud storage.

Usually human users users don't have writeAccess either to cloud storage so the `uploadObjectAsync` can simply throw if invalid response.